### ISSUE TITLE
security definitions

### ISSUE THREAD

**Issue by chandwki** [jhthorsen/mojolicious-plugin-openapi#36]
It looks like the plugin completely ignores security definitions. It would be really nice to be able to have an authentication method route to an auth controller and method.

For example,

```
securityDefinitions:
  AccessToken:
    x-mojo-to: 'user#token_auth'
    description: |
      For accessing the API a valid JWT token must be passed in the 'Authorization' header. A valid JWT token is generated by the /auth endpoint.

      The following syntax must be used in the 'Authorization' header:
        Bearer: xxxxxx.yyyyyyy.zzzzzz
    type: apiKey
    name: Authorization
    in: header
```
I know you can specify a single custom route to do authentication but the problem is that this is a catch all and doesn't support multiple authentication methods.

The problem I was running into is if I have a `/api/login` route that generates a JWT token obviously I don't want it to authenticate a JWT token that doesn't exist yet!

**Comment by jhthorsen** [comment]
Can you explain how this would work with code? What would "user#token_auth" look like? I don't think that would work. I think you need a helper method that is passed a callback or something...

Maybe this could work?

```
x-mojo-helper: "auth.token"
```

```
helper "auth.token" => sub {
  my ($c, $cb) = @_;
  # do stuff...
  $c->$cb(\@errors);
};
```

`@errors` could then be transformed into the error document. Empty list would mean success.

I like your idea initially, but I need to play around (or see some example code) to see if it feels right.

**Comment by leejo** [comment]
The spec says this is more descriptive than prescriptive:

> A declaration of the security schemes available to be used in the specification. This does not enforce the security schemes on the operations and only serves to provide the relevant details for each scheme.

Although then goes on to say it can be used in the endpoints. See the relevant [section of the spec](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#securityDefinitionsObject).

I could look at supporting this spec in the OAuth2::AuthorizaionServer plugin/module but at the moment I have absolutely no time to do this.

**Comment by mrenvoize** [comment]
I'll certainly try to wrap my head around this at some point and likely pitch into the OAuth2::AuthorizationServer side of it when I get a moment.

**Comment by jhthorsen** [comment]
@leejo: I think if we can wrap a sane API around this, then it would be awesome to take advantage of the existing spec.

**Comment by jhthorsen** [comment]
See #40

**Comment by jhthorsen** [comment]
I'm closing this, because I think it's been resolved in #40.
